File: Net9/040-059/044 CS Attributes/MyApp.cs
// 44 C# Attributes
// Essai d'utilisation des attributs en C#
//
// 2001-02-18   PV
// 2001-08-18   PV Beta2
// 2006-10-01	PV		VS2005
// 2012-02-25	PV		VS2010
// 2021-09-18	PV		VS2022, Net6
// 2023-01-10	PV		Net7
// 2023-11-18	PV		Net8 C#12
// 2024-11-15	PV		Net9 C#13

using System;
using static System.Console;

namespace CS044;

[AttributeUsage(AttributeTargets.Class)]
public class MonAttribut(int iVal): Attribute
{
    private readonly int iPriv = iVal;
    private string sInfo = "";

    public string Info
    {
        get => sInfo;
        set => sInfo = value;
    }

    public int IFlags => iPriv;
}

[MonAttribut(1, Info = "Info de MaClasse1")]
internal class MaClasse1
{
}

[MonAttribut(7)]
internal class MaClasse2
{
}

internal static class MyApp
{
    public static void Main()
    {
        MaClasse1 o1 = new();
        MaClasse2 o2 = new();

        Zap(o1);
        Zap(o2);

        //Console.ReadLine();
    }

    private static void Zap(object o)
    {
        var t = o.GetType();
        WriteLine(t.Name);

        var m = (MonAttribut)t.GetCustomAttributes(typeof(MonAttribut), false)[0];
        WriteLine("{0}, {1}", m.IFlags, m.Info);
    }
}

File: Net10/040-059/044 VB Attributes/MyApp.vb
' 44 VB Attributes
' Essai d'utilisation des attributs en VB
'
' 2001-02-18    PV
' 2001-08-18    PV Beta2
' 2006-10-01 	PV		VS2005
' 2012-02-25	PV		VS2010
' 2021-09-18 	PV		VS2022, Net6
' 2023-01-10	PV		Net7
' 2023-11-18	PV		Net8
' 2024-11-15	PV		Net9
' 2026-01-19	PV		Net10

<AttributeUsage(AttributeTargets.Class)> Public Class MyAttribute
    Inherits Attribute

    Public Sub New(iVal As Integer)
        Flags = iVal
        Info = ""
    End Sub

    Public Property Info As String

    Public ReadOnly Property Flags As Integer
End Class

<My(1, Info:="Info de MaClasse1")>
Friend Class MaClasse1

End Class

<My(7)>
Friend Class MaClasse2

End Class

Friend Class MyApp

    Public Shared Sub Main()
        Dim o1 As New MaClasse1()
        Dim o2 As New MaClasse2()

        Zap(o1)
        Zap(o2)

        'Console.ReadLine()
    End Sub

    Private Shared Sub Zap(o As Object)
        Dim tob As Type     ' Type de l'objet
        tob = o.GetType()
        WriteLine(tob.Name)

        Dim tat As Type     ' Type de l'attribut
        tat = GetType(MyAttribute)

        Dim m As MyAttribute
        m = CType(tob.GetCustomAttributes(tat, False)(0), MyAttribute)
        WriteLine("{0}, {1}", m.Flags, m.Info)
    End Sub

End Class
